tests/core/gui/test_main_window.py

```diff
@@ -0,0 +1,194 @@
+# Parsec Cloud (https://parsec.cloud) Copyright (c) AGPLv3 2019 Scille SAS
+
+import pytest
+from PyQt5 import QtCore, QtWidgets
+
+from parsec.core.gui.lang import translate
+from parsec.core.types import BackendOrganizationFileLinkAddr
+
+
+@pytest.fixture
+async def logged_gui_with_files(
+    aqtbot, running_backend, backend, autoclose_dialog, logged_gui, bob, monkeypatch
+):
+    w_w = await logged_gui.test_switch_to_workspaces_widget()
+
+    assert logged_gui.tab_center.count() == 1
+
+    monkeypatch.setattr(
+        "parsec.core.gui.workspaces_widget.get_text_input", lambda *args, **kwargs: ("w1")
+    )
+    monkeypatch.setattr(
+        "parsec.core.gui.files_widget.get_text_input", lambda *args, **kwargs: ("dir1")
+    )
+
+    await aqtbot.mouse_click(w_w.button_add_workspace, QtCore.Qt.LeftButton)
+
+    def workspace_button_ready():
+        assert w_w.layout_workspaces.count() == 1
+        wk_button = w_w.layout_workspaces.itemAt(0).widget()
+        assert not isinstance(wk_button, QtWidgets.QLabel)
+
+    await aqtbot.wait_until(workspace_button_ready)
+
+    f_w = logged_gui.test_get_files_widget()
+    wk_button = w_w.layout_workspaces.itemAt(0).widget()
+    async with aqtbot.wait_exposed(f_w), aqtbot.wait_signal(f_w.folder_changed):
+        await aqtbot.mouse_click(wk_button, QtCore.Qt.LeftButton)
+
+    def _entry_available():
+        assert f_w.workspace_fs.get_workspace_name() == "w1"
+        assert f_w.table_files.rowCount() == 1
+
+    await aqtbot.wait_until(_entry_available)
+
+    def folder_ready():
+        assert f_w.isVisible()
+        assert f_w.table_files.rowCount() == 2
+        folder = f_w.table_files.item(1, 1)
+        assert folder
+        assert folder.text() == "dir1"
+
+    async with aqtbot.wait_signal(f_w.folder_stat_success, timeout=3000):
+        await aqtbot.mouse_click(f_w.button_create_folder, QtCore.Qt.LeftButton)
+
+    await aqtbot.wait_until(folder_ready)
+
+    d_w = await logged_gui.test_switch_to_devices_widget()
+
+    def device_widget_ready():
+        assert d_w.isVisible()
+
+    await aqtbot.wait_until(device_widget_ready)
+
+    return logged_gui, w_w, f_w
+
+
+@pytest.mark.gui
+@pytest.mark.trio
+async def test_file_link(
+    aqtbot, running_backend, backend, autoclose_dialog, logged_gui_with_files, bob, monkeypatch
+):
+    logged_gui, w_w, f_w = logged_gui_with_files
+    url = BackendOrganizationFileLinkAddr.build(
+        f_w.core.device.organization_addr, f_w.workspace_fs.workspace_id, f_w.current_directory
+    )
+
+    class AvailableDevice:
```

Comment:
Not a fan of this kind of mock given it makes assumptions on the underlying implementation.

I think it would be better to create a fixture that returns a real `AvailableDevice` (which can be provided with a dummy file path for the moment to keep it implementation simple)

```python
@pytest.fixture
def alice_available_device(alice, tmpdir):
    key_file_path = Path(tmpdir) / "alice_device.key"
    # Given the fixture never provides the password, it's pretty pointless to
    # really save the device, hence I think we can just provide a dummy Path
    # and be gone with it ;-)
    _save_device_with_password(
    	key_file=key_file_path,
    	device=alice,
    	password="leJambonC'estTropBon"
    )
    return AvailableDevice(
	    key_file_path=key_file_path,
	    organization_id=alice.organization_id,
	    device_id=alice.device_id,
	    human_handle=alice.human_handle,
	    device_label=alice.device_label,
	    slug=alice.slug,
    )
```

Reply:
Thanks, it definitely better